weedcoco/repo/deposit.py

```diff
@@ -2,178 +2,362 @@
+            addition_hash
+        ):
+            raise ValidationError("There are identical images in the repository.")
+
+    def get_all_existing_hash(self, repository):
+        return {
+            os.path.splitext(path.split("/")[-1])[0]
+            for dataset in repository.datasets()
+            if dataset.identifier != self.identifier
+            for path in dataset.get_logical_paths()
+            if path.split("/")[0] == "images"
+        }
+
+    def create_image_hash(self):
+        self.image_hash = {
+            image_name: f"{get_image_hash(self.image_dir / image_name, 16)}{os.path.splitext(image_name)[-1]}"
+            for image_name in os.listdir(self.image_dir)
+            if check_if_approved_image_extension(image_name)
+        }
+
+    def retrieve_image_paths(self):
+        for root, directories, files in os.walk(self.image_dir):
+            for filename in files:
+                yield (os.path.join(root, filename), filename)
+
+    def compress_to_download(self, temp_dir):
+        mkdir_safely(temp_dir)
+        download_path = (temp_dir / self.identifier).with_suffix(".zip")
+        with tempfile.TemporaryDirectory() as tmpdir:
+            zip_path = pathlib.Path(tmpdir) / "bundle.zip"
+            with ZipFile(zip_path, "w") as zip:
+                zip.write(self.weedcoco_path, "weedcoco.json")
+                for image, image_name in self.retrieve_image_paths():
+                    zip.write(image, "images/" + image_name)
+            # XXX: this should be move() not copy(), but move resulted in files
+            #      that we could not delete or move in the Docker volume.
+            copy(zip_path, download_path)
+        return download_path
+
+
+class Repository:
+    """
+    Class representing the ocfl repository
+    ---
+    root (pathlib.Path): the root directory of the repository
+    disposition (str): the algorithm used by the ocfl library to map ids to paths
+    """
+
+    def __init__(self, root, disposition="pairtree"):
+        self.root = root
+        self.disposition = disposition
+        self._ocfl = None
+
+    def initialize(self):
+        if not self.root.is_dir():
+            # use a separate ocfl.Store because it will create an invalid
+            # ocfl_layout.json if we initialise with a disposition
+            ocfl_store = ocfl.Store(root=str(self.root))
+            ocfl_store.initialize()
+
+    @property
+    def ocfl(self):
+        if self._ocfl is not None:
+            return self._ocfl
+        self._ocfl = ocfl.Store(root=str(self.root), disposition=self.disposition)
+        return self._ocfl
+
+    def datasets(self):
+        self.ocfl.open_root_fs()
+        self.ocfl.check_root_structure()
+        for path in self.ocfl.object_paths():
+            yield RepositoryDataset(self, path.split("/")[-1])
+
+    def dataset(self, identifier):
+        return RepositoryDataset(self, identifier)
```

Comment:
Is it returning a new dataset with the constructor? Why it doesn't require the specific repo to return the dataset?

Reply:
`dataset` is a method on the Repository, so it's passing the repo in to the RepositoryDataset constructor as `self`.